tests/cpp_new/analysis/value_diff_test.cpp

```diff
@@ -0,0 +1,63 @@
+#include "gtest/gtest.h"
+
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/ir_builder.h"
+#include "taichi/ir/statements.h"
+
+namespace taichi {
+namespace lang {
+namespace irpass {
+namespace analysis {
+
+TEST(ValueDiffPtrIndex, ConstI32) {
+  IRBuilder builder;
+
+  auto *const1 = builder.get_int32(42);
+  auto *const2 = builder.get_int32(2);
+
+  const auto diff = value_diff_ptr_index(const1, const2);
+  EXPECT_TRUE(diff.is_diff_certain);
+  EXPECT_EQ(diff.diff_range, 40);
+}
+
+TEST(ValueDiffPtrIndex, ConstF32) {
+  IRBuilder builder;
+
+  auto *const1 = builder.get_float32(1.0f);
+  auto *const2 = builder.get_float32(1.0f);
+
+  const auto diff = value_diff_ptr_index(const1, const2);
+  EXPECT_FALSE(diff.is_diff_certain);
```

Comment:
We don't check floating-point numbers since the pass is for pointer indices. Maybe add some comment?

Reply:
Done